onnxruntime/core/providers/migraphx/gpu_data_transfer.cc

```diff
@@ -67,6 +67,9 @@ common::Status GPUDataTransfer::CopyTensorAsync(const Tensor& src, Tensor& dst,
     } else if (src_device.Type() == OrtDevice::GPU) {
       // copying between GPU, this is non-blocking
       HIP_CALL_THROW(hipMemcpyAsync(dst_data, src_data, bytes, hipMemcpyDeviceToDevice, static_cast<hipStream_t>(stream.GetHandle())));
+    } else {
+      // copy from other CPU memory to GPU, this is blocking
+      HIP_CALL_THROW(hipMemcpyWithStream(dst_data, src_data, bytes, hipMemcpyHostToDevice, static_cast<hipStream_t>(stream.GetHandle())));
```

Comment:
This code block is not needed since the copying cpu to gpu is handled in line 66.

Reply:
No, since the type of transfers here are different Above is with hipStreamAsync wherase the one here performs the transfer with the stream events on the GPU. 

Stream events are async but are gated within the hipruntime stream queue of events. Processing is gated based on the last event completing and this can potentially block. The IF case performs a copy and disregards  any other events  in the GPU queue before performing a copy and assumes you're targeting a region of memory that is not being operated on vs something that operates inplace.

Reply:
my understanding: there are 5 device types: CPU, GPU, FPGA, NPU and DML.
```
if dest is gpu:
    if src is cpu: # line 63
    else if src is gpu: # line 67
    else # this new branch handles src in [FPGA, NPU, DML] # line 70
```

Also, I suggest to use same implementation of gpu_data_transfer.cc on both rocm and migraphx.

Reply:
These are split up because we have use cases with MIGraphX EP only being built in or ROCm EP only being built in.

Reply:
Is there use cases to copy data from FPGA/NPU to GPU, otherwise that code will not have effect.